model/src/main/kotlin/utils/DatabaseUtils.kt

```diff
@@ -60,26 +67,28 @@ object DatabaseUtils {
             "Password for PostgreSQL storage is missing."
         }
 
-        val dataSourceConfig = HikariConfig().apply {
-            jdbcUrl = config.url
-            username = config.username
-            password = config.password
-            schema = config.schema
-            maximumPoolSize = maxPoolSize
-
-            val suffix = " - $applicationNameSuffix".takeIf { applicationNameSuffix.isNotEmpty() }.orEmpty()
-            addDataSourceProperty("ApplicationName", "$ORT_FULL_NAME$suffix")
-
-            // Configure SSL, see: https://jdbc.postgresql.org/documentation/head/connect.html
-            // Note that the "ssl" property is only a fallback in case "sslmode" is not used. Since we always set
-            // "sslmode", "ssl" is not required.
-            addDataSourceProperty("sslmode", config.sslmode)
-            addDataSourcePropertyIfDefined("sslcert", config.sslcert)
-            addDataSourcePropertyIfDefined("sslkey", config.sslkey)
-            addDataSourcePropertyIfDefined("sslrootcert", config.sslrootcert)
-        }
+        return dataSources.getOrElse(config) {
```

Comment:
Is the access to a `ConcurrentHashMap` not always thread-safe?

Reply:
I missed that you switched to `ConcurrentHashMap` from your previous PR. However, you should at least use `getOrPut()` instead `getOrElse()`, because for this there is a thread-safe implementation. But even with this, like with the current implementation, the passed lambda function could be executed twice for the same config, creating two data sources where only one of them ends up in the map. This should be fine, if creating the data source does not already create a connection.

Reply:
As Martin said, even if using `putIfAbsent()`, it can happen that multiple data sources are created. With the default configuration of Hikari, it eagerly populates the connection pool by creating the configured number of connections. So, to be on the safe side, you need to evaluate the return value of `putIfAbsent()` and close the newly created data source again if a non-*null* value is returned. See the [JavaDocs](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#putIfAbsent-K-V-).

Reply:
Or synchronize the whole code block and never create the duplicate data source.

Reply:
Decided to use `lazy` with a lambda function, which creates the `HikariDataSource` object. This way, even calling the lambda block of `getOrElse()` multiple times to the same time should not create a connection.